### ISSUE TITLE
[Bug] Resize will overwrite the file on disk without asking

### ISSUE THREAD

**Issue by RetroEdit** [WerWolv/ImHex#1548]
### Operating System

Windows

### What's the issue you encountered?

When applying the `Edit > Resize...` operation, this will automatically save to the file on disk. This is not intuitive to me, and seems hazardous if operations like this can overwrite files implicitly.

I do think it's plausible this is intentional and the expected default workflow, in which case this issue can be closed. In that case, I don't personally think this is a great default and wonder if there are global settings to prevent this.

### How can the issue be reproduced?

1) Open any file (I tested with both a 16-byte file and a 16.0 MiB file)
2) Click `Edit > Resize...` or right clicking and select `Resize...`
3) Enter a size different to the current size and click `Set`. (For my 16-byte file, I tested both 0x8, 0x10, and 0x16 -- 0x10 didn't seem to overwrite the file, but the other two did).
- The file in question will now be overwritten by the resized version.

Expected result: the operation would be applied to the buffered version of the file or otherwise delay saving changes until explicitly asked.

### ImHex Version

v1.32.2

### ImHex Build Type

- [ ] Nightly or built from sources

### Installation type

MSI

**Comment by WerWolv** [comment]
Hey
This is indeed the intended behavior for one main reason. ImHex has the ability to open arbitrarily big files. If we wanted to not immediately apply changes to the file on disk, we'd be forced to load the entire file into memory which isn't really an option for huge files.
People have suggested loading files into memory entirely when they're small enough but the issue there is how would we define a cut-off size? If we just set an arbitrary size (even if it's configurable), people will assume one thing and then be surprised that another thing is happening. Right now, the behavior is consistent for every file size.

Generally, ImHex allows you to undo these "destructive" operations and not let you lose data unless you close ImHex in-between so personally I don't believe it's that big of an issue.
If you have a suggestion though that doesn't just involve loading the entire file into memory, I'm happy to improve the system though

**Comment by RetroEdit** [comment]
There are other options besides writing to disk or loading the full file into memory. I've seen other hex editors store operations *as operations* that get applied on demand (i.e., when you load the relevant part of the file).

For resize, a plausible implementation seems to be have a file end address and then a trailing size with the fill-value. There are a few cases to deal with to ensure resizing smaller than larger is handled properly and doesn't access contents that would technically be truncated, but as long as zone delineation is handled properly, it could work.

For some operations like copy-pasting data, this may introduce layers of complexity to handle it efficiently.

**Comment by x4fx77x4f** [comment]
This issue and #1585 are duplicates of #841.
> This is indeed the intended behavior for one main reason. ImHex has the ability to open arbitrarily big files. If we wanted to not immediately apply changes to the file on disk, we'd be forced to load the entire file into memory which isn't really an option for huge files.

Saving changes immediately is a massive footgun. Do any other hex editors do that? I've personally unintentionally edited files because of this. The files that caused me to notice were luckily replaceable. I could've been screwed if I had done this to files that were one of a kind. Problems could've arisen if I hadn't noticed. How long has this been the case? I fear any file I've ever opened in ImHex since this misfeature was added might have been unintentionally edited.
> If we wanted to not immediately apply changes to the file on disk, we'd be forced to load the entire file into memory which isn't really an option for huge files. People have suggested loading files into memory entirely when they're small enough but the issue there is how would we define a cut-off size? If we just set an arbitrary size (even if it's configurable), people will assume one thing and then be surprised that another thing is happening. Right now, the behavior is consistent for every file size.

People are already assuming one thing and being surprised another thing is happening—or worse: not noticing; it took me 19 months to notice this behavior. ImHex should behave intuitively where practical; otherwise, show a modal warning the user of the hazard and asking them if they want to open the file anyway or cancel.

**Comment by lunathir** [comment]
So, rather than using the original file and making edits to that, maybe copy the file somewhere and make edits there, and once you’re done maybe copy the changed file and overwrite/save under a different name?

**Comment by WerWolv** [comment]
@lunathir Copying the file is not a decent option either if the file is very large.
I know some other tools do that but if I'm modifying a 100GB file, I do not want a tool to create a backup of that.

Also just had this very same discussion again on the Discord server so I'm not going to repeat myself here
![image](https://github.com/WerWolv/ImHex/assets/10835354/507aca9f-ad48-4250-978e-b5cdb0e0c3b2)
tldr; I can add a popup about it but, unless somebody here would like to dig into it themselves and PR an improvement, this is probably as far as I can reasonably go.

**Comment by WerWolv** [comment]
Latest commit now makes ImHex load files that are smaller than 128MB into memory. Everything that's larger than that will be opened as read-only and show a popup telling the user that any modification will go straight to the file and asking if they are sure about opening the file with read-write permissions.
Hope that improves things a bit.

**Comment by RetroEdit** [comment]
What about files that are *exactly* 128MB? 🙂 Jokes aside, I appreciate the improvement.
